File: examples/mcp-bridge/async-tool.go
package main

import (
	"encoding/json"
	"fmt"
	"time"

	"github.com/axisrobo/harmovela/aep"
	"github.com/axisrobo/harmovela/aep/bridge"
)

type sink struct {
	events []map[string]any
}

func (s *sink) Send(event map[string]any) error {
	s.events = append(s.events, event)
	return nil
}

func main() {
	events := &sink{}
	b := bridge.NewMcpBridge(events)

	b.RegisterTool(bridge.AsyncToolHandler("build", "build tool", func(args map[string]any, tracker *aep.TaskTracker) map[string]any {
		return map[string]any{"artifact": "app.bin"}
	}))

	resp := b.HandleRequest(map[string]any{
		"jsonrpc": "2.0", "id": 1, "method": "tools/call",
		"params": map[string]any{"name": "build", "arguments": map[string]any{"_task_id": "task_demo"}},
	})

	content, _ := json.Marshal(resp["result"].(map[string]any)["content"])
	fmt.Println("tools/call response:", string(content))

	time.Sleep(500 * time.Millisecond)

	types := make([]string, 0, len(events.events))
	for _, e := range events.events {
		types = append(types, fmt.Sprint(e["type"]))
	}
	fmt.Println("lifecycle events:", join(types, " → "))
}

func join(ss []string, sep string) string {
	r := ""
	for i, s := range ss {
		if i > 0 {
			r += sep
		}
		r += s
	}
	return r
}

File: examples/mcp-bridge/async-tool.java
package com.axisrobo.aep.examples;

import com.axisrobo.aep.McpBridge;
import com.fasterxml.jackson.databind.ObjectMapper;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;

/**
 * MCP bridge: register an async tool, call it, observe lifecycle events.
 * Run from implementations/java with the classpath set.
 */
public class McpBridgeDemo {
    private static final ObjectMapper MAPPER = new ObjectMapper();

    public static void main(String[] args) throws Exception {
        var events = new ArrayList<Map<String, Object>>();
        var bridge = new McpBridge((Map<String, Object> e) -> events.add(e));

        bridge.registerTool(McpBridge.asyncToolHandler("build", "build tool",
            (a, tracker) -> Map.of("artifact", "app.bin", "task_id", tracker.id())));

        var resp = bridge.handleRequest(Map.of(
            "jsonrpc", "2.0", "id", 1, "method", "tools/call",
            "params", Map.of("name", "build", "arguments", Map.of("_task_id", "task_demo"))
        ));
        @SuppressWarnings("unchecked")
        var content = (List<Map<String, Object>>) ((Map<String, Object>) resp.get("result")).get("content");
        System.out.println("tools/call response: " + MAPPER.writeValueAsString(content));

        Thread.sleep(500);

        var types = new ArrayList<String>();
        for (var e : events) types.add((String) e.get("type"));
        System.out.println("lifecycle events: " + String.join(" → ", types));
    }
}
